com.unity.render-pipelines.high-definition/Editor/VFXGraph/VFXHDRPSubTarget.cs

```diff
@@ -0,0 +1,387 @@
+                name = HDStructFields.FragInputs.name,
+                fields = fields.ToArray()
+            };
+
+            return fragInputs;
+        }
+
+        // A key difference between Material Shader and VFX Shader generation is how surface properties are provided. Material Shaders
+        // simply provide properties via UnityPerMaterial cbuffer. VFX expects these same properties to be computed in the vertex
+        // stage (because we must evaluate them with the VFX blocks), and packed with the interpolators for the fragment stage.
+        static StructDescriptor AppendVFXInterpolator(StructDescriptor interpolator, VFXContext context, VFXContextCompiledData contextData)
+        {
+            var fields = interpolator.fields.ToList();
+
+            var expressionToName = context.GetData().GetAttributes().ToDictionary(o => new VFXAttributeExpression(o.attrib) as VFXExpression, o => (new VFXAttributeExpression(o.attrib)).GetCodeString(null));
+            expressionToName = expressionToName.Union(contextData.uniformMapper.expressionToCode).ToDictionary(s => s.Key, s => s.Value);
+
+            var mainParameters = contextData.gpuMapper.CollectExpression(-1).ToArray();
+
+            int normalSemanticIndex = 0;
+
+            // Warning/TODO: FragmentParameters are created from the ShaderGraphVfxAsset.
+            // We may ultimately need to move this handling of VFX Interpolators + SurfaceDescriptionFunction function signature directly into the SG Generator (since it knows about the exposed properties).
+            foreach (string fragmentParameter in context.fragmentParameters)
+            {
+                var filteredNamedExpression = mainParameters.FirstOrDefault(o => fragmentParameter == o.name &&
+                    !(expressionToName.ContainsKey(o.exp) && expressionToName[o.exp] == o.name)); // if parameter already in the global scope, there's nothing to do
+
+                if (filteredNamedExpression.exp != null)
+                {
+                    var type = VFXExpression.TypeToType(filteredNamedExpression.exp.valueType);
+
+                    if (!kVFXShaderValueTypeyMap.TryGetValue(type, out var shaderValueType))
+                        continue;
+
+                    // TODO: NoInterpolation only for non-strips.
+                    var interpolationModifier = InterpolationModifier.NoInterpolation;
+
+                    fields.Add(new FieldDescriptor(HDStructFields.VaryingsMeshToPS.name, filteredNamedExpression.name, "", shaderValueType, $"NORMAL{normalSemanticIndex++}", "", StructFieldOptions.Static, interpolationModifier));
+                }
+            }
+
+            interpolator.fields = fields.ToArray();
+            return interpolator;
+        }
+
+        enum VFXAttributeType
+        {
+            Current,
+            Source
+        }
+
+        static string[] kVFXAttributeStructNames =
+        {
+            "Attributes",
+            "SourceAttributes"
+        };
```

Comment:
The "Attributes" naming can be confusing with the "builtin" struct field which has a completly different purpose : 
 https://github.com/Unity-Technologies/Graphics/blob/master/com.unity.shadergraph/Editor/Generation/TargetResources/StructFields.cs#L8
 Maybe we should prefix them.

Reply:
Yes we should prefix these or perhaps splice the contents directly into AttributesElement